packages/keyring-controller/src/KeyringController.test.ts

```diff
@@ -1392,3 +1563,57 @@ describe('KeyringController', () => {
     });
   });
 });
+
+type WithControllerCallback<ReturnValue> = ({
+  controller,
+  preferences,
+  initialState,
+}: {
+  controller: KeyringController;
+  preferences: {
+    setAccountLabel: sinon.SinonStub;
+    removeIdentity: sinon.SinonStub;
+    syncIdentities: sinon.SinonStub;
+    updateIdentities: sinon.SinonStub;
+    setSelectedAddress: sinon.SinonStub;
+  };
+  encryptor: MockEncryptor;
+  initialState: KeyringMemState;
+}) => Promise<ReturnValue> | ReturnValue;
+
+type WithControllerOptions = Partial<KeyringConfig>;
+
+type WithControllerArgs<ReturnValue> =
+  | [WithControllerCallback<ReturnValue>]
+  | [WithControllerOptions, WithControllerCallback<ReturnValue>];
+
+/**
+ * Builds a controller based on the given options, and calls the given function
+ * with that controller.
+ *
+ * @param args - Either a function, or an options bag + a function. The options
+ * bag is equivalent to the options that KeyringController takes;
+ * the function will be called with the built controller.
+ * @returns Whatever the callback returns.
+ */
+async function withController<ReturnValue>(
+  ...args: WithControllerArgs<ReturnValue>
+): Promise<ReturnValue> {
+  const [{ ...rest }, fn] = args.length === 2 ? args : [{}, args[0]];
+  const encryptor = new MockEncryptor();
+  const preferences = {
+    setAccountLabel: sinon.stub(),
+    removeIdentity: sinon.stub(),
+    syncIdentities: sinon.stub(),
+    updateIdentities: sinon.stub(),
+    setSelectedAddress: sinon.stub(),
+  };
+  const controller = new KeyringController(preferences, {
+    encryptor,
+    keyringBuilders: [keyringBuilderFactory(QRKeyring)],
```

Comment:
Is it intentional that this builds a QR keyring?

Reply:
This was already included in the `beforeEach` fn, it has been ported as is to `withController`.

I tried to remove it and the reason it was placed there is that `KeyringController.signTypedMessage` calls `getOrAddQRKeyring`, which needs the QR keyring builder to work properly. As this is the only method along with the QR ones to use that keyring builder, I will remove it from `withController` and will add another arg `additionalKeyrings` so that we can pass those from each test case when needed.